src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -140,9 +280,7 @@ private async Task<bool> IsEnabledAsync<TContext>(string feature, TContext appCo
 
                         //
                         // IFeatureFilter
-                        if (filter is IFeatureFilter featureFilter && await featureFilter
-                                                                                .EvaluateAsync(context, cancellationToken)
-                                                                                .ConfigureAwait(false))
+                        if (filter is IFeatureFilter featureFilter && await featureFilter.EvaluateAsync(context, cancellationToken).ConfigureAwait(false))
```

Comment:
Want to change this to the else clause of the `if(useAppContext)` too for consistency?

Reply:
I see. Looks like we didn't fail when a contextual filter is listed but without a context being provided today. This is a pseudo code I have in mind. Do you feel it's easier to understand the logic you explained above?

```
if (it's contextual filter)
{
    if (useAppContext)
    {
        do work for contextual filter
    }
    else
    {
        fail // we can skip this else block if we want to avoid the the breaking change.
    }    
}
else
{
    do work for non-contextual filter
}
```

I agree we should do the same for the assignment.

Reply:
I do think it makes it more clear that we expect a filter to be either contextual or noncontextual. Originally this was written with the design that a filter could be both, but we then added [a check](https://github.com/microsoft/FeatureManagement-Dotnet/blob/8d1774ff0521dbbf0f1f35551cc2d7115d1f65e9/src/Microsoft.FeatureManagement/FeatureManagementBuilder.cs#L35) to ensure that a filter isn't used that implements both.

Given that we only expect a filter to be one of the two types, your layout is probably better.

Reply:
The latest commit contains a change that is somewhat between your suggestion and what already existed. It does clarify that at the root we distinguish non-contextual filter vs. contextual filter. However, we do check if app context is being used before trying to check if a filter is a contextual filter since the app context type is used in that process.

Reply:
With the change these conditions result in exceptions, whereas they did not before.

if feature references a contextual filter and caller passes a different app context type then the contextual filter supports.
if feature references a contextual filter and caller doesn't pass an app context